libr/core/cmd_api.c

```diff
@@ -53,10 +90,39 @@ R_API RCmd *r_cmd_free(RCmd *cmd) {
 			R_FREE (cmd->cmds[i]);
 		}
 	}
+	free_cmd_desc_tree (cmd->root_cmd_desc);
 	free (cmd);
 	return NULL;
 }
 
+R_API RCmdDesc *r_cmd_get_root(RCmd *cmd) {
+	return cmd->root_cmd_desc;
+}
+
+static bool is_valid_desc(RCmdDesc *cd, const char *name) {
+	return cd->type == R_CMD_DESC_TYPE_OLDINPUT ||
+		(cd->type == R_CMD_DESC_TYPE_ARGV && !strcmp (cd->name, name));
+}
+
+R_API RCmdDesc *r_cmd_get_desc(RCmd *cmd, const char *cmd_identifier) {
+	r_return_val_if_fail (cmd && cmd_identifier, NULL);
+	char *cmdid = strdup (cmd_identifier);
+	char *end_cmdid = cmdid + strlen (cmdid);
+	RCmdDesc *res = NULL;
+	// match longer commands first
+	while (*cmdid) {
+		res = ht_pp_find (cmd->ht_cmds, cmdid, NULL);
+		r_warn_if_fail (!res || res->type == R_CMD_DESC_TYPE_OLDINPUT || res->type == R_CMD_DESC_TYPE_ARGV);
```

Comment:
Is this debug code? Wouldn't it print warning under normal use case where loop performs more than one iteration.

Reply:
it is "just to be sure". It shouldn't print anything in regular situation, it's only a kind of protection against me (the developer) from doing stupid thing

Reply:
I think I misunderstood the how warn_if_fail condition works.

Reply:
it's a sort of assert :) i just put it there to be sure to handle new cases as i will add new R_CMD_DESC_TYPE_*